File: src/main/java/com/protryon/jasm/LocalContext.java
package com.protryon.jasm;

import com.shapesecurity.functional.data.HashTable;
import com.shapesecurity.functional.data.Maybe;

public class LocalContext {

    public HashTable<Integer, Local> locals = HashTable.emptyUsingEquality();

    public Local getOrMakeLocal(int index) {
        Maybe<Local> localMaybe = locals.get(index);
        if (localMaybe.isJust()) {
            return localMaybe.fromJust();
        }
        Local newLocal = new Local(index);
        this.locals = this.locals.put(index, newLocal);
        return newLocal;
    }

    public Local getOrMakeLocal(int index, JType type) {
        Maybe<Local> localMaybe = locals.get(index);
        if (localMaybe.isJust()) {
            if (type != null && (localMaybe.fromJust().type == null || !type.assignableTo(localMaybe.fromJust().type))) {
                return this.updateLocal(localMaybe.fromJust().resetType(type));
            }
            return localMaybe.fromJust();
        }
        Local newLocal = new Local(index, type);
        this.locals = this.locals.put(index, newLocal);
        return newLocal;
    }

    public Local updateLocal(Local local) {
        this.locals = this.locals.put(local.index, local);
        return local;
    }

    public LocalContext fork() {
        LocalContext localContext = new LocalContext();
        localContext.locals = this.locals;
        return localContext;
    }

}

File: src/main/java/com/protryon/jasm/LocalContext.kt
package com.protryon.jasm

import com.shapesecurity.functional.data.HashTable
import com.shapesecurity.functional.data.Maybe

class LocalContext {

    var locals = HashTable.emptyUsingEquality<Int, Local>()

    fun getOrMakeLocal(index: Int): Local {
        val localMaybe = locals.get(index)
        if (localMaybe.isJust) {
            return localMaybe.fromJust()
        }
        val newLocal = Local(index)
        this.locals = this.locals.put(index, newLocal)
        return newLocal
    }

    fun getOrMakeLocal(index: Int, type: JType?): Local {
        val localMaybe = locals.get(index)
        if (localMaybe.isJust) {
            return if (type != null && (localMaybe.fromJust().type == null || !type.assignableTo(localMaybe.fromJust().type))) {
                this.updateLocal(localMaybe.fromJust().resetType(type))
            } else localMaybe.fromJust()
        }
        val newLocal = Local(index, type)
        this.locals = this.locals.put(index, newLocal)
        return newLocal
    }

    fun updateLocal(local: Local): Local {
        this.locals = this.locals.put(local.index, local)
        return local
    }

    fun fork(): LocalContext {
        val localContext = LocalContext()
        localContext.locals = this.locals
        return localContext
    }

}
